test/correctness/float16_t_neon_op_check.cpp

```diff
@@ -0,0 +1,377 @@
+        assert(tasks.size());
+        assert(tasks.back().op == op);
+        suffix_map.emplace(tasks.back().name, suffix);
+    }
+
+    void compile_and_check(Func error, const string &op, const string &name, int vector_width, std::ostringstream &error_msg) override {
+        std::string fn_name = "test_" + name;
+        std::string file_name = output_directory + fn_name;
+
+        auto ext = Internal::get_output_info(target);
+        std::map<Output, std::string> outputs = {
+            {Output::c_header, file_name + ext.at(Output::c_header).extension},
+            {Output::object, file_name + ext.at(Output::object).extension},
+            {Output::assembly, file_name + ".s"},
+        };
+        error.compile_to(outputs, arg_types, fn_name, target);
+
+        std::ifstream asm_file;
+        asm_file.open(file_name + ".s");
+
+        bool found_it = false;
+
+        string suffix = suffix_map[name];
+        std::ostringstream msg;
+        msg << op << " did not generate for target=" << target.to_string() << " suffix=" << suffix << " vector_width=" << vector_width << ". Instead we got:\n";
+
+        string line;
+        while (getline(asm_file, line)) {
+            msg << line << "\n";
+
+            // Check for the op in question. In addition, check if the expected suffix exists in the operand
+            found_it |= wildcard_search(op, line) && wildcard_search(suffix, line) && !wildcard_search("_" + op, line);
+        }
+
+        if (!found_it) {
+            error_msg << "Failed: " << msg.str() << "\n";
+        }
+
+        asm_file.close();
+    }
+
+    string suffix_of_st(int num_elements, int bits, int vector_size) {
+        assert(num_elements >= 2 && num_elements <= 4);
+        assert(vector_size % num_elements == 0);
+        const int num_lanes = vector_size / num_elements;
+        switch (bits) {
+        case 32:
+            return num_lanes == 2 ? ".2s" : ".4s";
+        case 16:
+            return num_lanes == 4 ? ".4h" : ".8h";
+        default:
+            assert(0);
+            return "unsupported_bits";
+        }
+    }
+
+    std::map<string, string> suffix_map;
+    const Var x{"x"}, y{"y"};
+};
+}  // namespace
+
+int main(int argc, char **argv) {
+    Target host = get_host_target();
+    Target hl_target = get_target_from_environment();
+    Target jit_target = get_jit_target_from_environment();
+    printf("host is:      %s\n", host.to_string().c_str());
+    printf("HL_TARGET is: %s\n", hl_target.to_string().c_str());
+    printf("HL_JIT_TARGET is: %s\n", jit_target.to_string().c_str());
+
+    // Only for 64bit target with fp16 feature
+    if (!(hl_target.arch == Target::ARM && hl_target.bits == 64 && hl_target.has_feature(Target::ARMFp16))) {
+        printf("[SKIP] To run this test, set HL_TARGET=arm-64-<os>-arm_fp16. \n");
+        return 0;
+    }
+    // Create Test Object
+    // Use smaller dimension than default(768, 128) to avoid fp16 overflow in reduction test case
```

Comment:
I know this is out of scope for this PR (and Halide...), but it would be great if there were fp16 -> fp32 reduction instructions, so we could efficiently reduce fp16 inputs to fp32 outputs.

Reply:
I agree. Thank you.

Reply:
(Let's open an issue for that so it doesn't get forgotten?)